### PR TITLE
core: adjust hardfork enabling logic

### LINKED ISSUE #3069: T5 testnet block 2418704 processing issue
blockQueue: failed adding block into the blockchain        {"error": "onPersist failed: VM has failed: at instruction 0 (SYSCALL): insufficient funds", "blockHeight": 2418703, "nextIndex": 2418704}

### LINKED ISSUE #3096: Port hardforks configuration structure changes
Port neo-project/neo#2885 fix.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Follow the logic described in https://github.com/neo-project/neo/pull/2886#issuecomment-1674745298 and port the https://github.com/neo-project/neo/pull/2886.

Close #3096.

### CODE REVIEW

**Comment by roman-khimov** [comment]
> #3069 

#3096!

File: pkg/config/hardfork.go [hunk 1]
```diff
@@ -17,12 +17,15 @@ const (
 	HFBasilisk // Basilisk
 )
 
+// Hardforks represents the ordered slice of all possible hardforks.
+var Hardforks = []Hardfork{HFAspidochelone, HFBasilisk}
```
Comment (roman-khimov) [inline pkg/config/hardfork.go:21]: Can we avoid repeating them?

File: pkg/config/protocol_config.go [hunk 2]
```diff
@@ -122,6 +122,24 @@ func (p *ProtocolConfiguration) Validate() error {
...
 		}
 	}
+	var (
+		prev             uint32
+		shouldBeDisabled bool
+	)
+	for _, cfgHf := range Hardforks {
+		h := p.Hardforks[cfgHf.String()]
+		if h != 0 && shouldBeDisabled {
+			return fmt.Errorf("hardfork %s should be disabled", cfgHf.String())
```
Comment (roman-khimov) [inline pkg/config/protocol_config.go:132]: It's a really unobvious way to say "missing previous hardfork configuration with cfgHf present`.

File: pkg/core/interop/context.go [hunk 3]
```diff
@@ -405,7 +405,15 @@ func (ic *Context) IsHardforkEnabled(hf config.Hardfork) bool {
 	if ok {
 		return ic.BlockHeight() >= height
 	}
-	return len(ic.Hardforks) == 0 // Enable each hard-fork by default.
+	if len(ic.Hardforks) == 0 {
+		return true // Enable each hard-fork by default if no hardforks specified in the config.
+	}
+	for _, cfgHf := range config.Hardforks {
```
Comment (roman-khimov) [inline pkg/core/interop/context.go:411]: Why don't we solve this in `NewBlockchain()` once (preparing a complete proper hf map for internal use) and not do these dances here?

File: pkg/config/hardfork.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/config/hardfork.go:21]: We can, but we need an ordered set of hardforks for validation checks simplicity. I've kept this list and removed the map, is it better?

File: pkg/config/protocol_config.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/config/protocol_config.go]: Fixed.

File: pkg/core/interop/context.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/core/interop/context.go]: Fixed.

File: pkg/config/hardfork.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/config/hardfork.go:21]: Please keep the map. What I mean is iterating over HF constants (from zero to the last+1 one) and filling this array in without copying names. So you wouldn't need to change this code once C one comes in.

Comment (AnnaShaleva) [reply pkg/config/hardfork.go]: Good, fixed.

Comment (AnnaShaleva) [reply pkg/config/hardfork.go]: However, we still to update the right bound in `for` loop each time new hardfork is added.

Comment (roman-khimov) [reply pkg/config/hardfork.go]: You can add unexprorted `hf_last` and do `i < hf_last`

Comment (AnnaShaleva) [reply pkg/config/hardfork.go]: See the updated code, like that?

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**